File: javascript/meetup/meetup.js
const Days = [ "Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];

function MeetupException(message){
  this.message = message;
  this.name = 'MeetupException';
}

function getDay( year, month, day ){
  return Days[new Date(year, month, day).getDay()];
}

function daysInMonth (month, year) {
  return new Date(year, month+1, 0).getDate()-1;
}

module.exports = function( year, month, day, week ) {
  if(week=="teenth"){
    for(let i=13; i<20; i++){
      if(getDay(year, month, i) == day){
        return new Date(year, month, i);
      }
    }
  } else if(week == "last") {
    let lastDay = daysInMonth( month, year );
    for(let i=lastDay+1; i>lastDay-7; i--){
      if(getDay(year, month, i) == day){
        return new Date(year, month, i);
      }
    }
  }

  let weekNo = parseInt(week.charAt(0))-1;
  let maxDay = parseInt(daysInMonth(month, year))
  let dayNo = 1;
  for(let i=0;i<5;i++) {
    for(let j=0; j<7; j++){
      if(i==weekNo){
        if(getDay(year, month, dayNo) == day){
          return new Date(year, month, dayNo);
        }
      }
      dayNo ++;
      if(dayNo>maxDay){ throw new MeetupException("This date is not valid"); return;}
    }
  }
}

File: python/meetup/meetup.py
from datetime import date

from calendar import monthrange

def meetup_day(year, month, day_of_the_week, which):
    dateDay = {}
    daysOfTheWeek = {}
    daysOfTheWeek["Monday"] = 0
    daysOfTheWeek["Tuesday"] = 1
    daysOfTheWeek["Wednesday"] = 2
    daysOfTheWeek["Thursday"] = 3
    daysOfTheWeek["Friday"] = 4
    daysOfTheWeek["Saturday"] = 5
    daysOfTheWeek["Sunday"] = 6
    day_of_the_week = daysOfTheWeek[day_of_the_week]
    for x in range(monthrange(year, month)[0]+1,monthrange(year, month)[1]-1):
        dateDay[x] = date(year,month,x).weekday()
    if(which == "1st" or which == "2nd" or which == "3rd" or which == "4th"):
        which = int(which[0])
        seen = 0
        x = monthrange(year, month)[0]
        while(seen!=which):
            if(dateDay[x]==day_of_the_week):
                seen+=1
                continue
            x+=1
        return date(year,month,x)
    elif(which=="teenth"):
            for x in range(10,20):
                if(dateDay[x]==day_of_the_week):
                    return date(year,month,x)
